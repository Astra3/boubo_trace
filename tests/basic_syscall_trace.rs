#[cfg(test)]
mod tests {
    use std::{process::Command, sync::Once};

    use boubo_trace::{
        syscall::{Syscall, SyscallIter, SyscallIterOpts, SyscallParseError},
        tracee::Tracee,
    };
    use insta::glob;
    use nix::unistd::Pid;
    use spawn_ptrace::CommandPtraceSpawn;
    static INIT: Once = Once::new();

    fn initialize() {
        INIT.call_once(|| {
            println!("running init");
            let mut cmd = Command::new("make")
                .current_dir("test_programs/")
                .spawn()
                .unwrap();
            cmd.wait().unwrap();
        });
    }

    #[test]
    fn simple_test() {
        initialize();
        glob!("../test_programs/build/", "*.exec", |exec| {
            println!("path: {}", exec.display());
            let mut cmd = Command::new(exec)
                .current_dir(exec.parent().unwrap())
                .spawn_ptrace()
                .unwrap();
            let pid = Pid::from_raw(cmd.id() as i32);

            let it = SyscallIter::new(Tracee::new(pid), SyscallIterOpts::default());
            let called_syscalls: Vec<_> = it
                .unwrap()
                .filter(|call| !matches!(call, Ok(Syscall::Unknown { .. })))
                .take_while(|call| !matches!(call, Err(SyscallParseError::ProcessExit(_))))
                .collect();
            insta::assert_debug_snapshot!(called_syscalls);
            cmd.wait().unwrap();
        });
    }
}
