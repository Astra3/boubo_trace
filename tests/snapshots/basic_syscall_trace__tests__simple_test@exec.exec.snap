---
source: tests/basic_syscall_trace.rs
expression: called_syscalls
input_file: test_programs/build/exec.exec
snapshot_kind: text
---
[
    Err(
        SyscallError {
            syscall: Execve,
            error: ENOENT,
        },
    ),
    Ok(
        Execve {
            pathname: [
                46,
                47,
                111,
                112,
                101,
                110,
                46,
                101,
                120,
                101,
                99,
                0,
            ],
            argv: [
                0,
            ],
            envp: [
                87,
                209,
                57,
                2,
                253,
                127,
                0,
            ],
        },
    ),
    Err(
        SyscallError {
            syscall: Openat,
            error: ENOENT,
        },
    ),
    Err(
        SyscallError {
            syscall: Openat,
            error: ENOENT,
        },
    ),
    Err(
        SyscallError {
            syscall: Openat,
            error: ENOENT,
        },
    ),
    Err(
        SyscallError {
            syscall: Openat,
            error: ENOENT,
        },
    ),
    Err(
        SyscallError {
            syscall: Openat,
            error: ENOENT,
        },
    ),
    Err(
        SyscallError {
            syscall: Openat,
            error: ENOENT,
        },
    ),
    Err(
        SyscallError {
            syscall: Openat,
            error: ENOENT,
        },
    ),
    Err(
        SyscallError {
            syscall: Openat,
            error: ENOENT,
        },
    ),
    Err(
        SyscallError {
            syscall: Openat,
            error: ENOENT,
        },
    ),
    Err(
        SyscallError {
            syscall: Openat,
            error: ENOENT,
        },
    ),
    Err(
        SyscallError {
            syscall: Openat,
            error: ENOENT,
        },
    ),
    Err(
        SyscallError {
            syscall: Openat,
            error: ENOENT,
        },
    ),
    Err(
        SyscallError {
            syscall: Openat,
            error: ENOENT,
        },
    ),
    Err(
        SyscallError {
            syscall: Openat,
            error: ENOENT,
        },
    ),
    Err(
        SyscallError {
            syscall: Openat,
            error: ENOENT,
        },
    ),
    Err(
        SyscallError {
            syscall: Openat,
            error: ENOENT,
        },
    ),
    Ok(
        Openat {
            dirfd: -100,
            pathname: [
                47,
                101,
                116,
                99,
                47,
                108,
                100,
                46,
                115,
                111,
                46,
                99,
                97,
                99,
                104,
                101,
                0,
            ],
            flags: OFlagSer(
                OFlag(
                    O_CLOEXEC,
                ),
            ),
            mode: ModeSer(
                Mode(
                    0x0,
                ),
            ),
        },
    ),
    Ok(
        Close {
            fd: 3,
        },
    ),
    Ok(
        Openat {
            dirfd: -100,
            pathname: [
                47,
                117,
                115,
                114,
                47,
                108,
                105,
                98,
                47,
                108,
                105,
                98,
                99,
                46,
                115,
                111,
                46,
                54,
                0,
            ],
            flags: OFlagSer(
                OFlag(
                    O_CLOEXEC,
                ),
            ),
            mode: ModeSer(
                Mode(
                    0x0,
                ),
            ),
        },
    ),
    Ok(
        Read {
            fd: 3,
            read_bytes: [
                127,
                69,
                76,
                70,
                2,
                1,
                1,
                3,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                3,
                0,
                62,
                0,
                1,
                0,
                0,
                0,
                96,
                118,
                2,
                0,
                0,
                0,
                0,
                0,
                64,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                120,
                173,
                30,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                64,
                0,
                56,
                0,
                15,
                0,
                64,
                0,
                63,
                0,
                62,
                0,
                6,
                0,
                0,
                0,
                4,
                0,
                0,
                0,
                64,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                64,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                64,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                72,
                3,
                0,
                0,
                0,
                0,
                0,
                0,
                72,
                3,
                0,
                0,
                0,
                0,
                0,
                0,
                8,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                3,
                0,
                0,
                0,
                4,
                0,
                0,
                0,
                160,
                50,
                30,
                0,
                0,
                0,
                0,
                0,
                160,
                50,
                30,
                0,
                0,
                0,
                0,
                0,
                160,
                50,
                30,
                0,
                0,
                0,
                0,
                0,
                30,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                30,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                16,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                1,
                0,
                0,
                0,
                4,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                104,
                51,
                2,
                0,
                0,
                0,
                0,
                0,
                104,
                51,
                2,
                0,
                0,
                0,
                0,
                0,
                0,
                16,
                0,
                0,
                0,
                0,
                0,
                0,
                1,
                0,
                0,
                0,
                5,
                0,
                0,
                0,
                0,
                64,
                2,
                0,
                0,
                0,
                0,
                0,
                0,
                64,
                2,
                0,
                0,
                0,
                0,
                0,
                0,
                64,
                2,
                0,
                0,
                0,
                0,
                0,
                217,
                1,
                23,
                0,
                0,
                0,
                0,
                0,
                217,
                1,
                23,
                0,
                0,
                0,
                0,
                0,
                0,
                16,
                0,
                0,
                0,
                0,
                0,
                0,
                1,
                0,
                0,
                0,
                4,
                0,
                0,
                0,
                0,
                80,
                25,
                0,
                0,
                0,
                0,
                0,
                0,
                80,
                25,
                0,
                0,
                0,
                0,
                0,
                0,
                80,
                25,
                0,
                0,
                0,
                0,
                0,
                190,
                226,
                4,
                0,
                0,
                0,
                0,
                0,
                190,
                226,
                4,
                0,
                0,
                0,
                0,
                0,
                0,
                16,
                0,
                0,
                0,
                0,
                0,
                0,
                1,
                0,
                0,
                0,
                6,
                0,
                0,
                0,
                120,
                59,
                30,
                0,
                0,
                0,
                0,
                0,
                120,
                75,
                30,
                0,
                0,
                0,
                0,
                0,
                120,
                75,
                30,
                0,
                0,
                0,
                0,
                0,
                80,
                75,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                209,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                16,
                0,
                0,
                0,
                0,
                0,
                0,
                2,
                0,
                0,
                0,
                6,
                0,
                0,
                0,
                64,
                105,
                30,
                0,
                0,
                0,
                0,
                0,
                64,
                121,
                30,
                0,
                0,
                0,
                0,
                0,
                64,
                121,
                30,
                0,
                0,
                0,
                0,
                0,
                48,
                2,
                0,
                0,
                0,
                0,
                0,
                0,
                48,
                2,
                0,
                0,
                0,
                0,
                0,
                0,
                8,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                4,
                0,
                0,
                0,
                4,
                0,
                0,
                0,
                136,
                3,
                0,
                0,
                0,
                0,
                0,
                0,
                136,
                3,
                0,
                0,
                0,
                0,
                0,
                0,
                136,
                3,
                0,
                0,
                0,
                0,
                0,
                0,
                80,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                80,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                8,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                4,
                0,
                0,
                0,
                4,
                0,
                0,
                0,
                216,
                3,
                0,
                0,
                0,
                0,
                0,
                0,
                216,
                3,
                0,
                0,
                0,
                0,
                0,
                0,
                216,
                3,
                0,
                0,
                0,
                0,
                0,
                0,
                36,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                36,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                4,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                4,
                0,
                0,
                0,
                4,
                0,
                0,
                0,
                24,
                50,
                30,
                0,
                0,
                0,
                0,
                0,
                24,
                50,
                30,
                0,
                0,
                0,
                0,
                0,
                24,
                50,
                30,
                0,
                0,
                0,
                0,
                0,
                32,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                32,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                4,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                7,
                0,
                0,
                0,
                4,
                0,
                0,
                0,
                120,
                59,
                30,
                0,
                0,
                0,
                0,
                0,
                120,
                75,
                30,
                0,
                0,
                0,
                0,
                0,
                120,
                75,
                30,
                0,
                0,
                0,
                0,
                0,
                16,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                136,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                8,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                83,
                229,
                116,
                100,
                4,
                0,
                0,
                0,
                136,
                3,
                0,
                0,
                0,
                0,
                0,
                0,
                136,
                3,
                0,
                0,
                0,
                0,
                0,
                0,
                136,
                3,
                0,
                0,
                0,
                0,
                0,
                0,
                80,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                80,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                8,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                80,
                229,
                116,
                100,
                4,
                0,
                0,
                0,
                160,
                175,
                27,
                0,
                0,
                0,
                0,
                0,
                160,
                175,
                27,
                0,
                0,
                0,
                0,
                0,
                160,
                175,
                27,
                0,
                0,
                0,
                0,
                0,
                204,
                121,
                0,
                0,
                0,
                0,
                0,
                0,
                204,
                121,
                0,
                0,
                0,
                0,
                0,
                0,
                4,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                81,
                229,
                116,
                100,
                6,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
            ],
            requested_count: 832,
        },
    ),
    Ok(
        Close {
            fd: 3,
        },
    ),
    Ok(
        Openat {
            dirfd: -100,
            pathname: [
                116,
                101,
                115,
                116,
                46,
                116,
                120,
                116,
                0,
            ],
            flags: OFlagSer(
                OFlag(
                    O_CREAT | O_RDWR | O_TRUNC,
                ),
            ),
            mode: ModeSer(
                Mode(
                    S_IRWXU,
                ),
            ),
        },
    ),
    Ok(
        Write {
            fd: 3,
            buf: [
                72,
                101,
                108,
                108,
                111,
                32,
                119,
                97,
                114,
                108,
                100,
                33,
                10,
            ],
        },
    ),
    Ok(
        Close {
            fd: 3,
        },
    ),
    Ok(
        ExitGroup {
            status: 0,
        },
    ),
]
